controllers/admin/AdminThemesController.php

```diff
@@ -789,31 +709,389 @@ private function translateAttributes($attributes)
+            );
+        }
+    }
+
+    /**
+     * Validate add access.
+     *
+     * If access is not granted, the thrown Exception's error message contains the translated rejection message.
+     *
+     * @param bool $allowDemoMode Should we grant access in demo mode ?
+     *
+     * @throws Exception
+     */
+    protected function validateAddAuthorization($allowDemoMode = false)
+    {
+        if ((_PS_MODE_DEMO_ && !$allowDemoMode)
+            || !$this->isAddGranted()
+        ) {
+            throw new Exception(
+                $this->trans('You do not have permission to add this.', array(), 'Admin.Notifications.Error')
+            );
+        }
+    }
+
+    /**
+     * Validate edit access.
+     *
+     * If access is not granted, the thrown Exception's error message contains the translated rejection message.
+     *
+     * @param bool $allowDemoMode Should we grant access in demo mode ?
+     *
+     * @throws Exception
+     */
+    protected function validateEditAuthorization($allowDemoMode = false)
+    {
+        if ((_PS_MODE_DEMO_ && !$allowDemoMode)
+            || !$this->isEditGranted()
+        ) {
+            throw new Exception(
+                $this->trans('You do not have permission to edit this.', array(), 'Admin.Notifications.Error')
+            );
+        }
+    }
+
+    /**
+     * Validate view access.
+     *
+     * If access is not granted, the thrown Exception's error message contains the translated rejection message.
+     *
+     * @param bool $allowDemoMode Should we grant access in demo mode ?
+     *
+     * @throws Exception
+     */
+    protected function validateViewAuthorization($allowDemoMode = false)
+    {
+        if ((_PS_MODE_DEMO_ && !$allowDemoMode)
+            || !$this->isViewGranted()
+        ) {
+            throw new Exception(
+                $this->trans('You do not have permission to view this.', array(), 'Admin.Notifications.Error')
+            );
+        }
+    }
+
+    protected function validateUploadSize()
+    {
+        $post_max_size = Tools::getMaxUploadSize();
+        if ($post_max_size
+            && isset($_SERVER['CONTENT_LENGTH'])
+            && $_SERVER['CONTENT_LENGTH']
+            && $_SERVER['CONTENT_LENGTH'] > $post_max_size
+        ) {
+            throw new Exception($this->trans('The uploaded file is too large.', array(), 'Admin.Design.Notification'));
+        }
+    }
+
+    /**
+     * Set redirect_after to same URL, but with an "&error" flag.
+     */
+    protected function addErrorToRedirectAfter()
+    {
+        $this->redirect_after = self::$currentIndex . '&token=' . $this->token . '&error';
+    }
+
+    /**
+     * Specific postProcess for "exporttheme" action
+     *
+     * @return bool false if access not granted
+     */
+    protected function postProcessExportTheme()
+    {
+        try {
+            $this->validateAddAuthorization(true);
```

Comment:
why do we need any specific access to export a theme? If its still required, validateEditAuthorization() seems to be better here.

Reply:
Already discussed with @vincentbz. He will be able to give you more details.